install/uninstall.ps1

```diff
@@ -0,0 +1,95 @@
+# Enable a default -Verbose flag for debug output.
+[CmdletBinding()]
+
+# Halt script execution at the first failed command.
+$script:ErrorActionPreference='Stop'
+
+$deleted = $false
+
+function Uninstall {
+    $uninstallDirectory = Get-UninstallDirectory
+    Remove-InstallationFolder $uninstallDirectory
+    Remove-FromPath $uninstallDirectory
+
+    if ($deleted) {
+        Write-Output "Uninstalled AzureAuth!"
+    } else {
+        Write-Output "There were no AzureAuth installations to be deleted."
+    }
+}
+
+function Get-UninstallDirectory {
+    $directory = if (![string]::IsNullOrEmpty($Env:AZUREAUTH_INSTALL_DIRECTORY)) {
+        Write-Error "Uninstalling from custom location is not supported"
+    } 
+    return ([System.IO.Path]::Combine($Env:LOCALAPPDATA, "Programs", "AzureAuth"))
+}
+
+function Remove-InstallationFolder {
+    param ([string]$directory)
+
+    if (Test-Path -Path $directory) {
+        Write-Verbose "Removing installations at '${directory}'"
+        Remove-Item -Force -Recurse $directory
+        $script:deleted = $true
+    } else {
+        Write-Verbose "There were no installations found at '${directory}'"
+    }
+}
+
+function Remove-FromPath {
+    param ([string]$uninstallDirectory)
+
+    $registryPath = 'Registry::HKEY_CURRENT_USER\Environment'
+    $currentPath = (Get-ItemProperty -Path $registryPath -Name PATH -ErrorAction SilentlyContinue).Path
+
+    # We try to find any other AzureAuth installations that are not in default location.
+    # Installations in custom locations that are not listed in PATH cannot be found.
+    $azureauthsInPath = (Get-Command -Name azureauth -ErrorAction SilentlyContinue -CommandType Application -All).Source
+    $otherDirectories = [System.Collections.ArrayList]@()
+    ForEach($az in $azureauthsInPath) {
+        if (!$az.Contains($uninstallDirectory)) {
+            $additionalDirectory = (Get-Item $az).Directory.FullName
+            Write-Verbose "Additional installation found in ${additionalDirectory}"
+            $_ = $otherDirectories.Add($additionalDirectory)
+            Remove-InstallationFolder ($additionalDirectory)

```

Comment:
From a separation of concerns perspective, this function becomes rather more complicated by doing this. At face value, by looking at the name `Remove-FromPath`, I expect this to only manipulate `$PATH`, not to _also_ manipulate the filesystem.

**If** we did go ahead with the notion of uninstalling from custom locations, I would consider reorganizing the logic like so

1. Discover **all** locations where AzureAuth might be installed (always include the default directory).
2. Then remove all directories in a single place in the script.
3. Then remove all related `$PATH` entries because they can't point to anything anymore.

The output of step 1 can be fed into step 2, which can then be fed into step 3.

I think I _now_ understand why `$deleted` is a global, but I don't think it would need to be if the steps were pipelined as above.

Reply:
I will go ahead with this redesign since now two people have mentioned it. I will ping request re-review once I'm done